### ISSUE TITLE
MatchingTracks does not match on entities

### ISSUE THREAD

**Issue by lcichanowicz** [linvi/tweetinvi#122]
It seems that `MatchedTweetReceivedEventArgs.MatchingTracks` only matches on tweet text.  However, a filtered stream can contain matching tweets that have no matching tweet text, because they match on entity values.  [(Per Twitter):](https://dev.twitter.com/streaming/overview/request-parameters#track)

> The `text` of the Tweet **_and some entity fields_** are considered for matches. Specifically, the text attribute of the Tweet, expanded_url and display_url for links and media, text for hashtags, and screen_name for user mentions are checked for matches.

Will TweetInvi match on entity values in a future release?

**Comment by linvi** [comment]
I will consider this. My main concern is a performance issue. At the current time I am in theory able to handle around 600 Millions tweets per day with an average Intel Core i5.

This seem to be a lot, but if a user does too much in the additional logic contained in the event handler `MatchingTweetReceived` this performance can quickly been reduced.

I might give an option `CheckAllEntities` set to `true` by default. Also I will give access to the `AnyTweetReceived` that will allow developers to have access to all the tweets received in the json but that were not matched by Tweetinvi.

**Comment by lcichanowicz** [comment]
I think developers sometimes become overly concerned with performance.  In this case, the performance concern is at the expense of the library behaving in an expected fashion.  I'd rather have 5.5 million properly matched tweets per day, than 6 million where half of them are useless, because they have no match--even though twitter sent them because they match.

If I `CreateFilteredStream` and `.AddTrack("myPhrase")`, _every_`.MatchingTweetReceived` `args` `.MatchingTracks.Count` should be at least one--never zero.  It seems nonsensical to have a `.MatchingTweetReceived` that does not match.  A match with no matches?  In other words, a "matching tweet" must have a match.

I'm fine with an option to turn off matching, for performance--great idea--but I'd turn it off entirely, not just on entities.  You either want matches or you don't.  I'd never want some matches but not all matches.  It would be nice to have the match have a `MatchSource` or `MatchPath` property: `tweet.text`, `entities.urls.expanded_url`, etc.

I'm not trying to sound harsh or critical.  I'm just speaking plainly, because this is the main thing I use your library for, and I have come to discover that the matching feature is misleading and incomplete.  Yikes.

**Comment by linvi** [comment]
Thanks for your comment. I will take this into consideration and will get back to you when I have.
I have decided to include this issue in version 0.9.11.

Thanks

**Comment by linvi** [comment]
Regarding the comment : 

> If I CreateFilteredStream and .AddTrack("myPhrase"), every.MatchingTweetReceived args .MatchingTracks.Count should be at least one--never zero. It seems nonsensical to have a .MatchingTweetReceived that does not match. A match with no matches? In other words, a "matching tweet" must have a match.

This should not happen and this is a bug. I've just looked at the matching process and I think the problem happens only for `StartStreamMatchingAnyConditionAsync` could you please confirm that this is the method that you are using.

Thanks

**Comment by linvi** [comment]
If you use the latest version of the source code (0.9.11.0) you will find that `FilteredStream` has been improved.

``` c#
var fs = Stream.CreateFilteredStream();

// If not set, StreamMatchType will match on everything.
fs.MatchType = StreamMatchType.TweetText | StreamMatchType.AllEntities;

fs.MatchingTweetReceived += (sender, args) =>
{
    var howWasItMatched = args.StreamMatchType;
};

fs.StartStreamMatchingAnyCondition();
```

Also if you have a suggestion regarding the naming of `StreamMatchType` I would be grateful. I cannot find a good naming for this.

Also, in the case where Twitter sent a Tweet but Tweetinvi was not able to match it, a new event is now raised.

``` c#
fs.NonMatchingTweetReceived += (sender, args) =>
{
    // This tweet was not matched by any criteria provided to tweetinvi
    // If you reduced the scope of the StreamMatchType please ensure that
    // it was correctly set up.
};
```

I hope this will help you regarding your problem.

**Comment by linvi** [comment]
I also added a new event `TweetReceived` that will be raised whether the filters have been matched or not. You will be able to check which type of filter have been matched again using the `args.StreamMatchType`.

**Comment by lcichanowicz** [comment]
I have always used `Tweetinvi.Streams.FilteredStream.StartStreamMatchingAnyCondition`; however, when I tried your alpha DLL, and it broke some of my code, I realized I was using 0.9.8.2 the whole time.  I apologize if I was logging issue that have since been resolved.  I guess I rarely use frequently-changing libraries.  I'm not in the habit of looking for updates.  I'll check frequently from now on.

I will do some experimentation with your latest build and report back later this week.

Thanks for the support.
